server/msteams/client.go

```diff
@@ -1120,3 +1164,99 @@ func GetAuthURL(redirectURL string, tenantID string, clientID string, clientSecr
 		oauth2.SetAuthURLParam("code_challenge", codeChallenge),
 	)
 }
+
+func (tc *ClientImpl) addMentions(message, teamID, channelID, chatID string) (string, []models.ChatMessageMentionable) {
+	specialMentions := map[string]bool{
+		"all":     true,
+		"channel": true,
+		"here":    true,
+	}
+
+	re := regexp.MustCompile(`@([a-z0-9.\-_]+)`)
+	channelMentions := re.FindAllString(message, -1)
+
+	mentions := []models.ChatMessageMentionable{}
+
+	for id, m := range channelMentions {
+		username := m[1:]
+		mentionedText := m
+
+		mention := models.NewChatMessageMention()
+		mentionedID := int32(id)
+		mention.SetId(&mentionedID)
+
+		mentioned := models.NewChatMessageMentionedIdentitySet()
+		conversation := models.NewTeamworkConversationIdentity()
+
+		if specialMentions[username] {
+			if chatID != "" {
+				chat, err := tc.GetChat(chatID)
+				if err != nil {
+					tc.plugin.LogDebug("Unable to get ms teams chat", "Error", err.Error())
+				} else {
+					if chat.Type == "G" {
+						mentionedText = "Everyone"
+					} else {
+						continue
+					}
+				}
+
+				conversation.SetId(&chatID)
+			} else {
+				msChannel, err := tc.GetChannel(teamID, channelID)
+				if err != nil {
+					tc.plugin.LogDebug("Unable to get ms teams channel", "Error", err.Error())
+				} else {
+					mentionedText = msChannel.DisplayName
+				}
+
+				conversation.SetId(&channelID)
+			}
+
+			conversation.SetDisplayName(&mentionedText)
+
+			conversationIdentityType := models.CHANNEL_TEAMWORKCONVERSATIONIDENTITYTYPE
+			conversation.SetConversationIdentityType(&conversationIdentityType)
+			mentioned.SetConversation(conversation)
+		} else {
+			mmUser, err := tc.plugin.GetUserByUsername(username)
```

Comment:
I have concerns about having to pass the whole API and the whole store to address this. That leads me to think that  it is not responsability of the client calculate the list of mentions. Maybe it should be done outside somehow.